frontends/p4/inlining.h

```diff
@@ -100,10 +100,52 @@ struct PerInstanceSubstitutions {
 struct InlineSummary : public IHasDbPrint {
     /// Various substitutions that must be applied for each instance
     struct PerCaller {  // information for each caller
+        /// 3-tuple identifying all the invocations of the subparser which can use the same
+        /// inlined states.
+        /// Consists of:
+        /// - Pointer to the declaration of the invoked subparser
+        /// - Pointer to the expression representing the transition statement from the state
+        ///   which invokes the subparser
+        /// - Pointer to a newly constructed IndexedVector of all parser statements
+        ///   which follow the subparser invocation (not including the transition statement)
+        typedef std::tuple<const IR::Declaration_Instance *,
+                           const IR::Expression *,
+                           const IR::IndexedVector<IR::StatOrDecl> *> InlinedInvocationInfo;
+
+        /// Hash for InlinedInvocationInfo used as a key for unordered_map
+        struct key_hash {
+            std::size_t operator() (const InlinedInvocationInfo &k) const {
+                std::ostringstream oss;
+                std::get<1>(k)->dbprint(oss);
+                for (auto e : *std::get<2>(k)) {
+                    e->dbprint(oss);
+                }
+                return std::hash<const IR::Declaration_Instance *>{}(std::get<0>(k)) ^
+                        std::hash<std::string>{}(oss.str());
+            }
+        };
+
+        /// Binary equality predicate for InlinedInvocationInfo used as a key for unordered_map
+        struct key_equal {
+            bool operator() (const InlinedInvocationInfo &v0,
```

Comment:
I can see several problems with this equality test:
- IR pointers are often cloned during traversal; it is not clear to me that you will be comparing the same version of an IR node
- I am not sure that equiv is the right test for expressions; two expressions can be syntactically equivalent but they can evaluate to different results *if they are invoked in different contexts*. For example, an expression that is just a variable 'a' will evaluate to a different result depending on where it is invoked.

Reply:
* If other maps in the same class (`declToCallee`, `substitutions`, `callToInstance`) can use pointer (`const IR::Declaration_Instance*`) I thought that I could rely on that it is not cloned in this pass and it should be all right to use it here as well. Or am I wrong?

* I think I see where you are heading, but I'm still not sure if that is an issue. If you have a variable 'a' it is not evaluated in compile time, but in run time, isn't it? So during compilation it is still just a variable 'a' and it may have different values during run time.
So f.e. these two parsers should be equivalent, shouldn't they?
```
parser Subparser(packet_in packet) {...}

parser ParserImpl(packet_in packet) {
  Subparser() p;
  bit<8> a;
  bit<8> b;

...

  state s1 {
    a = 1;
    p.apply(packet);
    b = a;
    transition accept;
  }

  state s2 {
    a = 2;
    p.apply(packet);
    b = a;
    transition accept;
  }
}
```
```
parser Subparser(packet_in packet) {...}

parser ParserImpl(packet_in packet) {
  Subparser() p;
  bit<8> a;
  bit<8> b;

...

  state s1 {
    a = 1;
    p.apply(packet);
    transition s3;
  }

  state s2 {
    a = 2;
    p.apply(packet);
    transition s3;
  }

  state s3 {
    b = a;
    transition accept;
  }
}
```
Could you please try to provide some counterexample which would be problematic?

Reply:
Can you also add a comment show the effect of this additional optimization on a small example?

Reply:
I checked that this is eliminated by UniqueNames and MoveDeclarations passes, which are run before Inline pass, so it should be all right. I added also a note in the comment that it relies on the fact that these passes are run before that.
I added also an additional test (subparser_inlining13.p4) which contains local variables declared in states.

I added also an example of the optimization in the comment.